_posts/2020-08-26-#19339.md

```diff
@@ -0,0 +1,94 @@
+In addition to enforcing consensus rules, it also applies the node’s policy. 
+[#19339](https://github.com/bitcoin/bitcoin/pull/19339) deals with acceptance logic that falls into neither category… and thus doesn’t belong in validation!
+
+* For reference, [Policy](https://github.com/bitcoin/bitcoin/tree/master/src/policy) = a node’s extra standards for validation, only applied to unconfirmed transactions. 
+It includes things like fees (e.g. [`-minrelaytxfee`](https://github.com/bitcoin/bitcoin/blob/3ab2582c7fe76d2839ab493512758d5601903c86/src/init.cpp#L547)) 
+and extra verification [flags](https://github.com/bitcoin/bitcoin/blob/3ab2582c7fe76d2839ab493512758d5601903c86/src/policy/policy.h#L56) (e.g. `SCRIPT_VERIFY_DISCOURAGE_UPGRADEABLE*`).
+This contrasts with consensus rules, which are universally applied in block validation.
+
+* On master, ATMP accepts an argument `const CAmount& nAbsurdFee`, and enforces it as a maximum fee amount; 
+this protects users from paying “absurdly high” fees resulting from an unintentional error. 
+It returns `TxValidationResult::TX_NOT_STANDARD` error for absurd fees, which seems to suggest that it is part of policy.
+
+* However, `nAbsurdFee` is only used for transactions originating from the node’s clients (RPC and wallet). 
+When a node receives a transaction from its peers, it uses an empty value for `nAbsurdFee` when calling ATMP
+(you can see the values of `nAbsurdFee` removed in [commit]()). 
+Furthermore, there is no way for a node operator to set the maximum fee, e.g. through a command line argument on start.
+
+* [#19339](https://github.com/bitcoin/bitcoin/pull/19339) redelegates fee checking from the mempool to the clients in order to make mempool logic less user-specific. 
+The main observable behavior change is in the error message, which is updated to include RPC and help reviewers confirm 
+the transition from validation logic to transaction error.
+
+  *  Note that this only changes RPCs [`testmempoolaccept`](https://github.com/bitcoin/bitcoin/blob/3ab2582c7fe76d2839ab493512758d5601903c86/src/rpc/rawtransaction.cpp#L858) 
+  and [`sendrawtransaction`](https://github.com/bitcoin/bitcoin/blob/3ab2582c7fe76d2839ab493512758d5601903c86/src/rpc/rawtransaction.cpp#L800).
+     There are no changes to wallet (reviewers may want to confirm that this is ok).
+     
+  * Note that clients often use [`BroadcastTransaction()`](https://github.com/bitcoin/bitcoin/blob/3ab2582c7fe76d2839ab493512758d5601903c86/src/node/transaction.cpp#L16), 
+    to interface with ATMP; it also takes `max_tx_fee` to pass in as `nAbsurdFee`. 
+    This PR adds fee checking to `BroadcastTransaction()` by adding a dry run of ATMP before the real submission.
+
+* To be careful to preserve the fee protection offered to the user, the PR includes a few intermediate stages. 
+The author recommends reviewing this PR by checking out each commit individually and confirming that the behavior is 
```

Comment:
s/checking out/reviewing/

usually there is no need to `git checkout` each commit separately to review the diff of each commit locally, unless one also wants to compile each commit to verify the build, run tests, etc.

Reply:
I am hoping they do that actually 😏  i spent so much time making sure each commit is solid hehe
but yeah, I will say "review" and they can determine for themselves if they want to go through the effort of checkout/build/test

Reply:
Hygienic commits are great. Personally, if it's a non-trivial PR I always debug-build at least at the head commit (which can catch warnings or errors), and sometimes I build/run tests at every commit if it's critical code or not obvious that each commit is standalone. I always review each commit separately and locally (I use gitk with dracula for the diff and also open the files in vim or emacs), but everyone has their review tools/process.

Reply:
Understood :D `abstract func review()`